packages/protocol/test/layout/layout.ts

```diff
@@ -0,0 +1,297 @@
+import { getBuildArtifacts } from '@openzeppelin/upgrades'
+
+/* HACK! truffle test was unable to compile this test (TypeScript would end up
+ * claiming that it couldn't find the name `assert`) without the following
+ * lines.
+ */
+import { MigrationsContract } from 'types'
+// @ts-ignore
+const Migrations: MigrationsContract = artifacts.require('Migrations')
+
+import { reportLayoutIncompatibilities } from '@celo/protocol/lib/layout'
+
+/* We store artifacts for the various test cases in ./test/resources/layout
+ * For each test case, there should be a build_<test case name> directory with
+ * truffle-compile output, and contracts_<test case name> directory with the
+ * corresponding contracts.
+ *
+ * The base for most of these tests is in contracts_original
+ */
+const getTestArtifacts = (caseName: string) => {
+  return getBuildArtifacts(`./test/resources/layout/build_${caseName}`)
+}
+
+const testCases = {
+  original: getTestArtifacts('original'),
+  inserted_constant: getTestArtifacts('inserted_constant'),
+  appended: getTestArtifacts('appended'),
+  inserted: getTestArtifacts('inserted'),
+  appended_in_parent: getTestArtifacts('appended_in_parent'),
+  removed: getTestArtifacts('removed'),
+  typechange: getTestArtifacts('typechange'),
+  typechange_in_struct: getTestArtifacts('typechange_in_struct'),
+  typechange_in_parent: getTestArtifacts('typechange_in_parent'),
+  typechange_in_library_struct: getTestArtifacts('typechange_in_library_struct'),
+  removed_from_struct: getTestArtifacts('removed_from_struct'),
+  removed_from_parent: getTestArtifacts('removed_from_parent'),
+  inserted_in_struct: getTestArtifacts('inserted_in_struct'),
+  inserted_in_library_struct: getTestArtifacts('inserted_in_library_struct'),
+  removed_from_library_struct: getTestArtifacts('removed_from_library_struct'),
+
+  original_complex: getTestArtifacts('original_complex'),
```

Comment:
Why have `original` and `original_complex`?

Reply:
I added tests with "complex" types (mappings and arrays) later. Rather than rewriting/recompiling all the previous tests I just created a fresh starting point for these tests.